src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -76,9 +87,78 @@ internal abstract class JsonPropertyInfo
         internal virtual void GetPolicies(JsonSerializerOptions options)
         {
             DetermineSerializationCapabilities(options);
+            DeterminePropertyName(options);
             IgnoreNullValues = options.IgnoreNullValues;
         }
 
+        private void DeterminePropertyName(JsonSerializerOptions options)
+        {
+            if (PropertyInfo != null)
+            {
+                JsonPropertyNameAttribute nameAttribute = GetAttribute<JsonPropertyNameAttribute>();
+                if (nameAttribute != null)
+                {
+                    NameAsString = nameAttribute.Name;
+                    
+                    // This is detected and thrown by caller.
+                    if (NameAsString == null)
+                    {
+                        return;
+                    }
+                }
+                else if (options.PropertyNamingPolicy != null)
+                {
+                    NameAsString = options.PropertyNamingPolicy.ConvertName(PropertyInfo.Name);
+
+                    // This is detected and thrown by caller.
+                    if (NameAsString == null)
+                    {
+                        return;
+                    }
+                }
+                else
+                {
+                    NameAsString = PropertyInfo.Name;
+                }
+
+                // At this point propertyName is valid UTF16, so just call the simple UTF16->UTF8 encoder.
+                _name = Encoding.UTF8.GetBytes(NameAsString);
+
+                // Set the compare name.
+                if (options.PropertyNameCaseInsensitive)
+                {
+                    CompareNameAsString = NameAsString.ToUpperInvariant();
```

Comment:
nit: `Compare` is a verb. What are the semantics here. Can we rename this? `NameUsedForComparison`?